src/librustc/mir/mod.rs

```diff
@@ -2998,7 +2997,7 @@ pub struct UnsafetyCheckResult {
 /// The layout of generator state
 #[derive(Clone, Debug, RustcEncodable, RustcDecodable, HashStable)]
 pub struct GeneratorLayout<'tcx> {
-    pub fields: Vec<LocalDecl<'tcx>>,
+    pub variant_fields: IndexVec<VariantIdx, Vec<LocalDecl<'tcx>>>,
```

Comment:
I think the inner `Vec` could be `IndexVec<Field`.
Also, surprised this is `LocalDecl`, is that only for debuginfo?
If so, can you add something similar to `__upvar_debuginfo_codegen_only_do_not_use`, to ensure there is no cross-talk between debuginfo and non-debuginfo usecases?

Reply:
Hmm, no, we also use it to get the type of the fields in `GeneratorSubsts::state_tys()`. I suppose I can turn it into a struct with something like `__upvar_debuginfo_codegen_only_do_not_use` and whatever struct we should be using here. (What should we be using? Right now we only need the type, I think.)

Reply:
I meant that `GeneratorLayout` should *only* have `Ty<'tcx>`, and that debuginfo should be separate.

Reply:
I separated these out and made a `GeneratorField` newtype. That way our layout code will also know when the same field is in two variants so it can lay them out correctly (see the FIXME I added).